### PR TITLE
Add HttpAccessTokenRefreshError

### PR DESCRIPTION
**Pull request by thobrla**
This adds an exception class including the HTTP status for access
token refresh requests made over HTTP.

Fixes https://github.com/google/oauth2client/issues/309

### CODE REVIEW

File: tests/test_client.py [hunk 1]
```diff
@@ -695,9 +695,9 @@ def test_token_refresh_failure(self):
             http = self.credentials.authorize(http)
             try:
                 http.request('http://example.com')
-                self.fail('should raise AccessTokenRefreshError exception')
-            except AccessTokenRefreshError:
-                pass
+                self.fail('should raise HttpAccessTokenRefreshError exception')
+            except HttpAccessTokenRefreshError, e:
+                self.assertEqual(400, e.status)
```
Comment (nathanielmanistaatgoogle) [inline tests/test_client.py]: http_client.BAD_REQUEST rather than the literal value 400, please.

File: oauth2client/client.py [hunk 2]
```diff
@@ -134,6 +134,13 @@ class AccessTokenRefreshError(Error):
     """Error trying to refresh an expired access token."""
 
 
+class HttpAccessTokenRefreshError(AccessTokenRefreshError):
+    """Error (with HTTP status) trying to refresh an expired access token."""
+    def __init__(self, *args, **kwargs):
+        super(HttpAccessTokenRefreshError, self).__init__(*args)
```
Comment (nathanielmanistaatgoogle) [inline oauth2client/client.py:140]: What happens to keyword args that do not have "status" as their key? It looks like they get dropped on the floor?

Comment (thobrla) [inline oauth2client/client.py:140]: I'm not aware that the base exception class does anything with keyword arguments to begin with, so I don't think forwarding them to the base class is a problem.  If you disagree, I'm find with including them.

File: tests/test_client.py (hunk 1 above)
Comment (thobrla) [inline tests/test_client.py]: Done.

File: oauth2client/client.py (hunk 2 above)
Comment (dhermes) [inline oauth2client/client.py:140]: It [inherits](https://github.com/google/oauth2client/blob/29347a91356a958f500b2c59ae3ed85387c282bb/oauth2client/client.py#L125-L134) via `AccessTokenRefreshError -> Error -> Exception` and neither subclass actually modifies the parent (beyond the name). I can confirm that `Exception` rejects `kwargs`:

``` python
>>> Exception(a=10)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: exceptions.Exception does not take keyword arguments
```

---

Unfortunately, if you want to pass `*args`, you use named `kwargs` so this is the only way to go.

**Comment by dhermes** [comment]
Also `except HttpAccessTokenRefreshError, e:` is not acceptable syntax in Python 3. See Travis [failure][1]

[1]: https://travis-ci.org/google/oauth2client/jobs/81469436

File: tests/test_client.py [hunk 3]
```diff
@@ -690,14 +691,15 @@ def test_token_refresh_failure(self):
...
+                 b'{"error":"access_denied"}'),
             ])
             http = self.credentials.authorize(http)
             try:
                 http.request('http://example.com')
-                self.fail('should raise AccessTokenRefreshError exception')
-            except AccessTokenRefreshError:
-                pass
+                self.fail('should raise HttpAccessTokenRefreshError exception')
+            except HttpAccessTokenRefreshError, e:
```
Comment (dhermes) [inline tests/test_client.py]: Using `except HttpAccessTokenRefreshError as e:`

File: tests/test_client.py [hunk 4]
```diff
@@ -690,14 +691,15 @@ def test_token_refresh_failure(self):
...
+                ({'status': http_client.BAD_REQUEST},
+                 b'{"error":"access_denied"}'),
             ])
             http = self.credentials.authorize(http)
             try:
                 http.request('http://example.com')
-                self.fail('should raise AccessTokenRefreshError exception')
-            except AccessTokenRefreshError:
-                pass
+                self.fail('should raise HttpAccessTokenRefreshError exception')
```
Comment (dhermes) [inline tests/test_client.py:700]: While you're mucking with this code, mind switching from `self.fail` and catching the error to just using the modern approach: `self.assertRaises`?

Comment (thobrla) [inline tests/test_client.py:700]: I don't see a clean way to use assertRaises and also check the status property of the exception.  Do you have a suggestion here?

Comment (dhermes) [inline tests/test_client.py:700]: You're right, I forgot about that part.

**Comment by dhermes** [comment]
All LGTM, but before merging I am going to parrot for @nathanielmanistaatgoogle 

He [usually asks][1] people

> Please squash commits with each round of code review back-and-forth; the successive stages of progress here shouldn't necessarily become a part of canonical project history.

He [also asks][2] that you

> Please rewrite your commit message to conform to http://chris.beams.io/posts/git-commit/#seven-rules

[1]: https://github.com/google/oauth2client/pull/185#issuecomment-103598795
[2]: https://github.com/google/oauth2client/pull/196#issuecomment-112982411

**Comment by thobrla** [comment]
Yes, was waiting for LGTM to squash.  In theory, the initial commit message should follow those rules (the referenced issue explains the "why").

**Comment by dhermes** [comment]
@thobrla How desperate are you for a release? We did `1.5.0` on 2015-09-03 and `1.5.1` on 2015-09-16 (previously the release schedule is much less frequest).

**Comment by thobrla** [comment]
Not desperate.  I'll need it for gsutil 4.16 but that release does not yet have a date planned.  I can pin the git submodule to this commit until then.

**Comment by dhermes** [comment]
OK. Sounds like we can be independent then?

Side note: Aren't you guys moving to Go for the CLI?

**Comment by thobrla** [comment]
Not completely independent - I will need an oauth2client release prior to gsutil 4.16, but if one hasn't happened by the time that is ready, I will ask for it.

**Comment by dhermes** [comment]
OK SGTM.

As for Golang?

**Pull request merged**